File: app/src/main/java/mk/webfactory/template/feature/common/ui/BasePresenter.java
package mk.webfactory.template.feature.common.ui;

import static mk.webfactory.template.util.Preconditions.checkArgument;
import static mk.webfactory.template.util.Preconditions.checkNotNull;
import static mk.webfactory.template.util.Preconditions.checkState;

/**
 * @deprecated - We'll use view model
 */
@Deprecated
public abstract class BasePresenter<View> {

  private View view;

  public final void takeView(View view) {
    checkState(this.view == null, "takeView before previous view is dropped.");
    this.view = checkNotNull(view, "view == null");
    onTakeView(view);
  }

  public final void dropView(View view) {
    checkArgument(view != null, "dropView with null argument.");
    if (this.view != view) { return; }
    onDropView();
    this.view = null;
  }

  protected abstract void onTakeView(View v);

  protected abstract void onDropView();

  public final boolean hasView() {
    return view != null;
  }

  protected final View getView() {
    return view;
  }
}

File: app/src/main/java/mk/webfactory/template/feature/common/ui/BasePresenter.kt
/*
 * MIT License
 *
 * Copyright (c) 2020 Web Factory LLC
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package mk.webfactory.template.feature.common.ui

/**
 * Base presenter for the MVP pattern.
 */
abstract class BasePresenter<View> {

    protected var view: View? = null
        private set

    fun takeView(view: View) {
        if (this.view != null) throw IllegalStateException("takeView before previous view is dropped.")
        this.view = view
        onTakeView(view)
    }

    fun dropView(view: View) {
        if (this.view !== view) {
            return
        }
        onDropView()
        this.view = null
    }

    protected abstract fun onTakeView(v: View)

    protected abstract fun onDropView()

    fun hasView(): Boolean {
        return view != null
    }
}